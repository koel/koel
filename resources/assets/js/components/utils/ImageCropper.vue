<template>
  <div class="w-full h-full fixed top-0 left-0 flex items-center justify-center z-[99] bg-black/70">
    <div class="relative max-w-full max-h-full rounded-md flex">
      <Cropper
        ref="cropper"
        :max-width="config.maxWidth"
        :min-width="config.minWidth"
        :src="source"
        :stencil-props="{ aspectRatio: 1 }"
      />
      <div class="fixed top-6 right-6 flex flex-1 gap-2">
        <Btn success @click.prevent="crop">Crop</Btn>
        <Btn transparent @click.prevent="emits('cancel')">Cancel</Btn>
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref, toRefs } from 'vue'
import { Cropper } from 'vue-advanced-cropper'
import 'vue-advanced-cropper/dist/style.css'

import Btn from '@/components/ui/form/Btn.vue'

const props = withDefaults(defineProps<{
  source?: string | null
  config?: {
    minWidth: number
    maxWidth?: number
  }
}>(), {
  source: null,
  config: () => ({
    minWidth: 192,
  }),
})

const emits = defineEmits<{
  (e: 'crop', result: string): void
  (e: 'cancel'): void
}>()
const { source, config } = toRefs(props)
const cropper = ref<typeof Cropper>()

const crop = () => emits('crop', cropper.value?.getResult().canvas.toDataURL())
</script>
