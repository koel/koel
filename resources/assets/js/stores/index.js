export * from "./album";
export * from "./artist";
export * from "./equalizer";
export * from "./favorite";
export * from "./playlist";
export * from "./preference";
export * from "./queue";
export * from "./setting";
export * from "./shared";
export * from "./song";
export * from "./user";
