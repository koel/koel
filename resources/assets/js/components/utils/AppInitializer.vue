<template>
  <slot />
</template>

<script lang="ts" setup>
import { onMounted } from 'vue'
import { useErrorHandler, useOverlay } from '@/composables'
import { commonStore, preferenceStore as preferences } from '@/stores'
import { socketListener, socketService, uploadService } from '@/services'

const emits = defineEmits<{
  (e: 'success'): void
  (e: 'error', err: unknown): void
}>()

const { showOverlay, hideOverlay } = useOverlay()

/**
 * Request for notification permission if it's not provided and the user is OK with notifications.
 */
const requestNotificationPermission = async () => {
  if (preferences.show_now_playing_notification && window.Notification && window.Notification.permission !== 'granted') {
    preferences.show_now_playing_notification = await window.Notification.requestPermission() === 'denied'
  }
}

onMounted(async () => {
  showOverlay({ message: 'Just a little patienceâ€¦' })

  try {
    await commonStore.init()

    await requestNotificationPermission()

    window.addEventListener('beforeunload', (e: BeforeUnloadEvent) => {
      if (uploadService.shouldWarnUponWindowUnload() || preferences.confirm_before_closing) {
        e.preventDefault()
        e.returnValue = ''
      }
    })

    await socketService.init() && socketListener.listen()

    emits('success')
  } catch (error: unknown) {
    useErrorHandler().handleHttpError(error)
    emits('error', error)
  } finally {
    hideOverlay()
  }
})
</script>

<style lang="postcss" scoped>

</style>
