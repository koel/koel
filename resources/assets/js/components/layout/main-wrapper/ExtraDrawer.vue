<template>
  <aside :class="{ 'showing-pane': activeTab }">
    <div class="controls">
      <div class="top">
        <SidebarMenuToggleButton class="burger" />
        <ExtraDrawerTabHeader v-if="song" v-model="activeTab" />
      </div>

      <div class="bottom">
        <button
          v-koel-tooltip.left
          :title="shouldNotifyNewVersion ? 'New version available!' : 'About Koel'"
          type="button"
          @click.prevent="openAboutKoelModal"
        >
          <icon :icon="faInfoCircle" />
          <span v-if="shouldNotifyNewVersion" class="new-version-notification" />
        </button>

        <button v-koel-tooltip.left title="Log out" type="button" @click.prevent="logout">
          <icon :icon="faArrowRightFromBracket" />
        </button>

        <ProfileAvatar @click="onProfileLinkClick" />
      </div>
    </div>

    <div v-if="song" v-show="activeTab" class="panes">
      <div
        v-show="activeTab === 'Lyrics'"
        id="extraPanelLyrics"
        aria-labelledby="extraTabLyrics"
        role="tabpanel"
        tabindex="0"
      >
        <LyricsPane :song="song" />
      </div>

      <div
        v-show="activeTab === 'Artist'"
        id="extraPanelArtist"
        aria-labelledby="extraTabArtist"
        role="tabpanel"
        tabindex="0"
      >
        <ArtistInfo v-if="artist" :artist="artist" mode="aside" />
        <span v-else>Loading…</span>
      </div>

      <div
        v-show="activeTab === 'Album'"
        id="extraPanelAlbum"
        aria-labelledby="extraTabAlbum"
        role="tabpanel"
        tabindex="0"
      >
        <AlbumInfo v-if="album" :album="album" mode="aside" />
        <span v-else>Loading…</span>
      </div>

      <div
        v-show="activeTab === 'YouTube'"
        id="extraPanelYouTube"
        data-testid="extra-drawer-youtube"
        aria-labelledby="extraTabYouTube"
        role="tabpanel"
        tabindex="0"
      >
        <YouTubeVideoList v-if="useYouTube && song" :song="song" />
      </div>
    </div>
  </aside>
</template>

<script lang="ts" setup>
import isMobile from 'ismobilejs'
import { faArrowRightFromBracket, faInfoCircle } from '@fortawesome/free-solid-svg-icons'
import { defineAsyncComponent, onMounted, ref, watch } from 'vue'
import { albumStore, artistStore, preferenceStore } from '@/stores'
import { useAuthorization, useNewVersionNotification, useThirdPartyServices } from '@/composables'
import { eventBus, logger, requireInjection } from '@/utils'
import { CurrentSongKey } from '@/symbols'

import ProfileAvatar from '@/components/ui/ProfileAvatar.vue'
import SidebarMenuToggleButton from '@/components/ui/SidebarMenuToggleButton.vue'

const LyricsPane = defineAsyncComponent(() => import('@/components/ui/LyricsPane.vue'))
const ArtistInfo = defineAsyncComponent(() => import('@/components/artist/ArtistInfo.vue'))
const AlbumInfo = defineAsyncComponent(() => import('@/components/album/AlbumInfo.vue'))
const YouTubeVideoList = defineAsyncComponent(() => import('@/components/ui/YouTubeVideoList.vue'))
const ExtraDrawerTabHeader = defineAsyncComponent(() => import('@/components/ui/ExtraDrawerTabHeader.vue'))

const { currentUser } = useAuthorization()
const { useYouTube } = useThirdPartyServices()
const { shouldNotifyNewVersion } = useNewVersionNotification()

const song = requireInjection(CurrentSongKey, ref(null))
const activeTab = ref<ExtraPanelTab | null>(null)

const artist = ref<Artist>()
const album = ref<Album>()

watch(song, song => song && fetchSongInfo(song))
watch(activeTab, tab => (preferenceStore.activeExtraPanelTab = tab))

const fetchSongInfo = async (_song: Song) => {
  song.value = _song
  artist.value = undefined
  album.value = undefined

  try {
    artist.value = await artistStore.resolve(_song.artist_id)
    album.value = await albumStore.resolve(_song.album_id)
  } catch (error) {
    logger.log('Failed to fetch media information', error)
  }
}

const openAboutKoelModal = () => eventBus.emit('MODAL_SHOW_ABOUT_KOEL')
const onProfileLinkClick = () => isMobile.any && (activeTab.value = null)
const logout = () => eventBus.emit('LOG_OUT')

onMounted(() => isMobile.any || (activeTab.value = preferenceStore.activeExtraPanelTab))
</script>

<style lang="scss" scoped>
aside {
  display: flex;
  flex-direction: row-reverse;
  color: var(--color-text-secondary);
  height: var(--header-height);
  z-index: 2;

  @media screen and (max-width: 768px) {
    @include themed-background();
    flex-direction: column;
    position: fixed;
    top: 0;
    width: 100%;

    &.showing-pane {
      height: 100%;
    }
  }
}

.panes {
  width: var(--extra-drawer-width);
  padding: 2rem 1.7rem;
  background: var(--color-bg-secondary);
  overflow: auto;
  box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.1);

  @media (hover: none) {
    // Enable scroll with momentum on touch devices
    overflow-y: scroll;
    -webkit-overflow-scrolling: touch;
  }

  @media screen and (max-width: 768px) {
    width: 100%;
    height: calc(100vh - var(--header-height) - var(--footer-height));
  }
}

.controls {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  height: 100%;
  width: 64px;
  padding: 1.6rem 0 1.2rem;
  background-color: rgba(0, 0, 0, .05);
  border-left: 1px solid rgba(255, 255, 255, .05);

  @media screen and (max-width: 768px) {
    z-index: 2;
    height: auto;
    width: 100%;
    flex-direction: row;
    padding: .5rem 1rem;
    border-bottom: 1px solid rgba(255, 255, 255, .05);
    box-shadow: 0 0 30px 0 rgba(0, 0, 0, .5);
  }

  .top, .bottom {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;
    gap: 1rem;

    @media screen and (max-width: 768px) {
      flex-direction: row;
      gap: .25rem;
    }
  }

  :deep(button) {
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
    height: 42px;
    aspect-ratio: 1/1;
    border-radius: 999rem;
    background: rgba(0, 0, 0, .3);
    font-size: 1.2rem;
    opacity: .7;
    transition: opacity .2s ease-in-out;
    color: currentColor;
    cursor: pointer;

    @media screen and (max-width: 768px) {
      background: none;
    }

    &:hover, &.active {
      opacity: 1;
      color: var(--color-text-primary);
    }

    &:active {
      transform: scale(.9);
    }

    &.burger {
      display: none;

      @media screen and (max-width: 768px) {
        display: block;
      }
    }
  }

  .new-version-notification {
    position: absolute;
    width: 10px;
    height: 10px;
    background: var(--color-highlight);
    right: 1px;
    top: 1px;
    border-radius: 50%;
  }
}
</style>
