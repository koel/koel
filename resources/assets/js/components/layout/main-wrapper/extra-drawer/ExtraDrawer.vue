<template>
  <aside
    :class="{ 'showing-pane': activeTab }"
    class="fixed sm:relative top-0 w-screen md:w-auto flex flex-col md:flex-row-reverse z-[2] text-k-text-secondary"
  >
    <div
      class="controls flex md:flex-col justify-between items-center md:w-[64px] md:py-6 tw:px-0
      bg-black/5 md:border-l border-solid md:border-l-white/5 md:border-b-0 md:shadow-none
      z-[2] w-screen flex-row border-b border-b-white/5 border-l-0 shadow-xl
      py-0 px-6 h-k-header-height"
    >
      <div class="btn-group">
        <ExtraDrawerButton class="md:hidden" @click.prevent="expandSidebar">
          <Icon :icon="faBars" fixed-width />
        </ExtraDrawerButton>
        <ExtraDrawerTabHeader v-if="songPlaying" v-model="activeTab" />
      </div>

      <div class="btn-group">
        <AboutKoelButton />
        <LogoutButton />
        <ProfileAvatar @click="onProfileLinkClick" />
      </div>
    </div>

    <div v-if="songPlaying" v-show="activeTab" class="panes py-8 px-6 overflow-auto bg-k-bg-secondary">
      <div
        v-show="activeTab === 'Lyrics'"
        id="extraPanelLyrics"
        aria-labelledby="extraTabLyrics"
        role="tabpanel"
        tabindex="0"
      >
        <LyricsPane v-if="playable" :song="playable" />
      </div>

      <div
        v-show="activeTab === 'Artist'"
        id="extraPanelArtist"
        aria-labelledby="extraTabArtist"
        role="tabpanel"
        tabindex="0"
      >
        <ArtistInfo v-if="artist" :artist="artist" mode="aside" />
        <span v-else>Loading…</span>
      </div>

      <div
        v-show="activeTab === 'Album'"
        id="extraPanelAlbum"
        aria-labelledby="extraTabAlbum"
        role="tabpanel"
        tabindex="0"
      >
        <AlbumInfo v-if="album" :album="album" mode="aside" />
        <span v-else>Loading…</span>
      </div>

      <div
        v-show="activeTab === 'YouTube'"
        id="extraPanelYouTube"
        aria-labelledby="extraTabYouTube"
        data-testid="extra-drawer-youtube"
        role="tabpanel"
        tabindex="0"
      >
        <YouTubeVideoList v-if="shouldShowYouTubeTab && playable" :song="playable" />
      </div>
    </div>
  </aside>
</template>

<script lang="ts" setup>
import isMobile from 'ismobilejs'
import { faBars } from '@fortawesome/free-solid-svg-icons'
import type { Ref } from 'vue'
import { computed, defineAsyncComponent, onMounted, ref, watch } from 'vue'
import { albumStore, artistStore, preferenceStore } from '@/stores'
import { useErrorHandler, useThirdPartyServices } from '@/composables'
import { eventBus, isSong, requireInjection } from '@/utils'
import { CurrentPlayableKey } from '@/symbols'

import ProfileAvatar from '@/components/ui/ProfileAvatar.vue'
import AboutKoelButton from '@/components/layout/main-wrapper/extra-drawer/AboutKoelButton.vue'
import LogoutButton from '@/components/layout/main-wrapper/extra-drawer/LogoutButton.vue'
import ExtraDrawerButton from '@/components/layout/main-wrapper/extra-drawer/ExtraDrawerButton.vue'

const LyricsPane = defineAsyncComponent(() => import('@/components/ui/LyricsPane.vue'))
const ArtistInfo = defineAsyncComponent(() => import('@/components/artist/ArtistInfo.vue'))
const AlbumInfo = defineAsyncComponent(() => import('@/components/album/AlbumInfo.vue'))
const YouTubeVideoList = defineAsyncComponent(() => import('@/components/ui/youtube/YouTubeVideoList.vue'))
const ExtraDrawerTabHeader = defineAsyncComponent(() => import('./ExtraDrawerTabHeader.vue'))

const { useYouTube } = useThirdPartyServices()

const playable = requireInjection(CurrentPlayableKey, ref(undefined)) as Ref<Song | undefined>
const activeTab = ref<ExtraPanelTab | null>(null)

const artist = ref<Artist>()
const album = ref<Album>()

const songPlaying = computed(() => playable.value && isSong(playable.value))
const shouldShowYouTubeTab = computed(() => useYouTube.value && songPlaying.value)

const fetchSongInfo = async (song: Song) => {
  playable.value = song
  artist.value = undefined
  album.value = undefined

  try {
    artist.value = await artistStore.resolve(song.artist_id)
    album.value = await albumStore.resolve(song.album_id)
  } catch (error: unknown) {
    useErrorHandler().handleHttpError(error)
  }
}

watch(playable, song => {
  if (!song || !isSong(song)) {
    return
  }
  fetchSongInfo(song)
}, { immediate: true })

watch(activeTab, tab => (preferenceStore.active_extra_panel_tab = tab))

const onProfileLinkClick = () => isMobile.any && (activeTab.value = null)
const expandSidebar = () => eventBus.emit('TOGGLE_SIDEBAR')

onMounted(() => isMobile.any || (activeTab.value = preferenceStore.active_extra_panel_tab))
</script>

<style lang="postcss" scoped>
@import '@/../css/partials/mixins.pcss';

@tailwind utilities;

@layer utilities {
  .btn-group {
    @apply flex md:flex-col justify-between items-center gap-1 md:gap-3;
  }
}

aside {
  @media screen and (max-width: 768px) {
    @mixin themed-background;

    &.showing-pane {
      height: 100%;
    }
  }
}

.panes {
  @apply no-hover:overflow-y-auto w-k-extra-drawer-width;

  box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.1);

  @media screen and (max-width: 768px) {
    width: 100%;
    height: calc(100vh - var(--header-height) - var(--footer-height));
  }
}
</style>
