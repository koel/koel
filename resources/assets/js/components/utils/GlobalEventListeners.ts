/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */

import { defineComponent, onMounted } from 'vue'
import { authService } from '@/services'
import { playlistFolderStore, playlistStore, userStore } from '@/stores'
import { eventBus, forceReloadWindow } from '@/utils'
import { useDialogBox, useMessageToaster, useRouter } from '@/composables'

export const GlobalEventListeners = defineComponent({
  setup (props, { slots }) {
    let toastSuccess: ReturnType<typeof useMessageToaster>['toastSuccess']
    let showConfirmDialog: ReturnType<typeof useDialogBox>['showConfirmDialog']
    let go: ReturnType<typeof useRouter>['go']

    onMounted(() => {
      toastSuccess = useMessageToaster().toastSuccess
      showConfirmDialog = useDialogBox().showConfirmDialog
      go = useRouter().go
    })

    eventBus.on('PLAYLIST_DELETE', async playlist => {
      if (await showConfirmDialog(`Delete the playlist "${playlist.name}"?`)) {
        await playlistStore.delete(playlist)
        toastSuccess(`Playlist "${playlist.name}" deleted.`)
        go('home')
      }
    }).on('PLAYLIST_FOLDER_DELETE', async folder => {
      if (await showConfirmDialog(`Delete the playlist folder "${folder.name}"?`)) {
        await playlistFolderStore.delete(folder)
        toastSuccess(`Playlist folder "${folder.name}" deleted.`)
        go('home')
      }
    }).on('LOG_OUT', async () => {
      await userStore.logout()
      authService.destroy()
      forceReloadWindow()
    })

    return () => slots.default?.()
  }
})
