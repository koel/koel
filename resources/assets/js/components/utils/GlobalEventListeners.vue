<template>
  <span></span>
</template>

<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import { authService } from '@/services'
import { playlistFolderStore, playlistStore, preferenceStore, userStore } from '@/stores'
import { eventBus, forceReloadWindow, requireInjection } from '@/utils'
import { DialogBoxKey, MessageToasterKey, RouterKey } from '@/symbols'

const toaster = requireInjection(MessageToasterKey)
const dialog = requireInjection(DialogBoxKey)
const router = requireInjection(RouterKey)

eventBus.on({
  PLAYLIST_DELETE: async (playlist: Playlist) => {
    if (await dialog.value.confirm(`Delete the playlist "${playlist.name}"?`)) {
      await playlistStore.delete(playlist)
      toaster.value.success(`Playlist "${playlist.name}" deleted.`)
      router.go('home')
    }
  },

  PLAYLIST_FOLDER_DELETE: async (folder: PlaylistFolder) => {
    if (await dialog.value.confirm(`Delete the playlist folder "${folder.name}"?`)) {
      await playlistFolderStore.delete(folder)
      toaster.value.success(`Playlist folder "${folder.name}" deleted.`)
      router.go('home')
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  LOG_OUT: async () => {
    await userStore.logout()
    authService.destroy()
    forceReloadWindow()
  }
})

router.onRouteChanged(() => {
  // Hide the extra panel away if a main view is triggered on mobile.
  if (isMobile.phone) {
    preferenceStore.showExtraPanel = false
  }
})
</script>
