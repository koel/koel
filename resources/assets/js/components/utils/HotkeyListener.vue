<template>
  <GlobalEvents
    @keydown.space="togglePlayback"
    @keydown.j="playNext"
    @keydown.k="playPrev"
    @keydown.f="search"
    @keydown.l="toggleLike"
  />
</template>

<script lang="ts" setup>
import { GlobalEvents } from 'vue-global-events'
import { eventBus } from '@/utils'
import { playbackService, socketService } from '@/services'
import { favoriteStore, queueStore } from '@/stores'

const togglePlayback = (e: KeyboardEvent) => {
  if (
    !(e.target instanceof Document) &&
    (e.target as Element).matches('input, textarea, button, select') &&
    !(e.target as Element).matches('input[type=range]')
  ) {
    return true
  }

  e.preventDefault()
  playbackService.toggle()

  return false
}

/**
 * Play the previous song when user presses K.
 */
const playPrev = (e: KeyboardEvent) => {
  if (!(e.target instanceof Document) && (e.target as Element).matches('input, select, textarea')) {
    return true
  }

  e.preventDefault()
  playbackService.playPrev()

  return false
}

/**
 * Play the next song when user presses J.
 */
const playNext = (e: KeyboardEvent) => {
  if (!(e.target instanceof Document) && (e.target as Element).matches('input, select, textarea')) {
    return true
  }

  e.preventDefault()
  playbackService.playNext()

  return false
}

/**
 * Put focus into the search field when user presses F.
 */
const search = (e: KeyboardEvent) => {
  if (
    !(e.target instanceof Document) &&
    (e.target as Element).matches('input, select, textarea') && !(e.target as Element).matches('input[type=range]')
  ) {
    return true
  }

  if (e.metaKey || e.ctrlKey) {
    return true
  }

  e.preventDefault()
  eventBus.emit('FOCUS_SEARCH_FIELD')

  return false
}

/**
 * Like/unlike the current song when use presses L.
 */
const toggleLike = (e: KeyboardEvent) => {
  if (!(e.target instanceof Document) && (e.target as Element).matches('input, select, textarea')) {
    return true
  }

  if (!queueStore.current) {
    return false
  }

  favoriteStore.toggleOne(queueStore.current)
  socketService.broadcast('SOCKET_SONG', queueStore.current)

  return false
}
</script>
