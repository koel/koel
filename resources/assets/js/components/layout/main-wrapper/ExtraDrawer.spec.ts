import { ref, Ref } from 'vue'
import { expect, it } from 'vitest'
import { screen, waitFor } from '@testing-library/vue'
import factory from '@/__tests__/factory'
import { albumStore, artistStore, commonStore, preferenceStore } from '@/stores'
import UnitTestCase from '@/__tests__/UnitTestCase'
import { CurrentSongKey } from '@/symbols'
import { eventBus } from '@/utils'
import ExtraDrawer from './ExtraDrawer.vue'

new class extends UnitTestCase {
  private renderComponent (songRef: Ref<Song | null> = ref(null)) {
    return this.render(ExtraDrawer, {
      global: {
        stubs: {
          ProfileAvatar: this.stub(),
          LyricsPane: this.stub('lyrics'),
          AlbumInfo: this.stub('album-info'),
          ArtistInfo: this.stub('artist-info'),
          YouTubeVideoList: this.stub('youtube-video-list'),
          ExtraPanelTabHeader: this.stub()
        },
        provide: {
          [<symbol>CurrentSongKey]: songRef
        }
      }
    })
  }

  protected test () {
    it('renders without a current song', () => expect(this.renderComponent().html()).toMatchSnapshot())

    it('sets the active tab to the preference', async () => {
      preferenceStore.activeExtraPanelTab = 'YouTube'
      this.renderComponent(ref(factory<Song>('song')))
      const tab = screen.getByTestId<HTMLElement>('extra-drawer-youtube')

      expect(tab.style.display).toBe('none')
      await this.tick()
      expect(tab.style.display).toBe('')
    })

    it('fetches info for the current song', async () => {
      commonStore.state.use_you_tube = true
      const artist = factory<Artist>('artist')
      const resolveArtistMock = this.mock(artistStore, 'resolve').mockResolvedValue(artist)

      const album = factory<Album>('album')
      const resolveAlbumMock = this.mock(albumStore, 'resolve').mockResolvedValue(album)

      const song = factory<Song>('song')

      const songRef = ref<Song | null>(null)

      this.renderComponent(songRef)
      songRef.value = song

      await waitFor(() => {
        expect(resolveArtistMock).toHaveBeenCalledWith(song.artist_id)
        expect(resolveAlbumMock).toHaveBeenCalledWith(song.album_id)
        ;['lyrics', 'album-info', 'artist-info', 'youtube-video-list'].forEach(id => screen.getByTestId(id))
      })
    })

    it('shows About Koel model', async () => {
      const emitMock = this.mock(eventBus, 'emit')
      this.renderComponent()

      await this.user.click(screen.getByRole('button', { name: 'About Koel' }))

      expect(emitMock).toHaveBeenCalledWith('MODAL_SHOW_ABOUT_KOEL')
    })

    it('notifies new version', async () => {
      it('shows new version', () => {
        commonStore.state.current_version = 'v1.0.0'
        commonStore.state.latest_version = 'v1.0.1'
        this.actingAsAdmin().renderComponent().getByRole('button', { name: 'New version available!' })
      })
    })

    it('logs out', async () => {
      const emitMock = this.mock(eventBus, 'emit')
      this.renderComponent()

      await this.user.click(screen.getByRole('button', { name: 'Log out' }))

      expect(emitMock).toHaveBeenCalledWith('LOG_OUT')
    })
  }
}
