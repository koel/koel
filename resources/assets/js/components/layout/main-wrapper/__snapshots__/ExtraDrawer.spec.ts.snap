// Vitest Snapshot v1

exports[`renders without a current song 1`] = `
<aside class="" data-v-119c862c="">
  <div class="controls" data-v-119c862c="">
    <div class="top" data-v-119c862c=""><button class="burger" data-v-119c862c=""><br data-testid="icon" icon="[object Object]"></button>
      <!--v-if-->
    </div>
    <div class="bottom" data-v-119c862c=""><button title="About Koel" type="button" data-v-119c862c=""><br data-testid="icon" icon="[object Object]" data-v-119c862c="">
        <!--v-if-->
      </button><button title="Log out" type="button" data-v-119c862c=""><br data-testid="icon" icon="[object Object]" data-v-119c862c=""></button><br data-testid="stub" data-v-119c862c=""></div>
  </div>
  <!--v-if-->
</aside>
`;
